server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java

```diff
@@ -237,7 +236,9 @@ public synchronized boolean awaitClose(long timeout, TimeUnit unit) throws Inter
         if (bulkRequest.numberOfActions() > 0) {
             execute();
         }
-        return this.bulkRequestHandler.awaitClose(timeout, unit);
+        boolean awaitClose = this.bulkRequestHandler.awaitClose(timeout, unit);
+        onClose.run();
```

Comment:
I'm wondering if we should ensure that the onClose.run() is always executed (like in a finally block) if something goes wrong in bulkRequestHandler.awaitClose(). I don't have a strong feeling about this, and if we decide to add it we should also have a test for it, but I think it would be safer.